### PR TITLE
Rewrite for RN 0.50+

### PR DESCRIPTION
**Pull request by jasonfma**
Major upgrade that supports 0.50+
- support by adding `reactNativeVersion` and `DeviceInfo` properties to the `NativeModules`
- Remove `Navigator` test as it was removed from RN
- Fix `BackAndroid` and add `BackHandler` tests
- update all packages
- fix lint issues
- drop support for Node 4

NOTE: This is a PR into the existing `rewrite` branch

### CODE REVIEW

**Comment by jasonfma** [comment]
Changing to support RN 0.50+ because the `AndroidProgressBar` had a breaking change in RN 0.49 that was fixed in RN 0.50. Rather than start off backwards compatible with 0.49, just do 0.50+.

**Comment by jasonfma** [comment]
Dropping support for Node 4. https://travis-ci.org/RealOrangeOne/react-native-mock/jobs/297785954

**Comment by dusave** [comment]
Should probably add `react`, `react-dom`, and `react-native` as dev dependencies also. When trying to build locally without those packages, it fails

**Comment by RealOrangeOne** [comment]
Initial skim looks good. Would probably be ideal to update the `README` stating that we've dropped support for Node `<6` (and why?). Will give it a proper review soon, but glad to see the tests are passing!

**Comment by gabceb** [comment]
Looking forward to this @RealOrangeOne and @jasonfma. When I try to pull this branch as a dev dependency for my project running RN 0.50 and React 16 it complains when running haste.js as part of the post install

```
Error: EISDIR: illegal operation on a directory, read
    at Error (native)
    at Object.fs.readSync (fs.js:731:19)
    at tryReadSync (fs.js:486:20)
    at Object.fs.readFileSync (fs.js:534:19)
    at /Users/gabrielcebrian/Projects/tes/mobile/app-job-mobile/node_modules/react-native-mock/src/haste.js:32:39
    at Function._.each._.forEach (/Users/gabrielcebrian/Projects/tes/mobile/app-job-mobile/node_modules/underscore/underscore.js:153:9)
    at Object.<anonymous> (/Users/gabrielcebrian/Projects/tes/mobile/app-job-mobile/node_modules/react-native-mock/src/haste.js:31:3)
    at Module._compile (module.js:570:32)
    at Object.Module._extensions..js (module.js:579:10)
    at Module.load (module.js:487:32)
```

If I clone the project on a different folder the post-install finishes as expected

**Comment by EdenGottlieb** [comment]
Adding a mock for FlatList would be a great idea, in my opinion.

File: .travis.yml [hunk 1]
```diff
@@ -1,22 +1,16 @@
 language: node_js
 node_js:
-  - "4.0.0"
   - "6.0.0"
-  - "4"
   - "6"
+  - "8"
```
Comment (RealOrangeOne) [inline .travis.yml]: Could we keep the base versions here too, so add `8.0.0`

File: tests/integration/apis/Easing.test.js [hunk 2]
```diff
@@ -14,7 +14,7 @@ import { expect } from 'chai';
 describe('Easing', () => {
   const { Easing } = require('react-native');
 
-  it('should have all functions', function () {
+  it('should have all functions', () => {
```
Comment (RealOrangeOne) [inline tests/integration/apis/Easing.test.js:17]: Please revert changes like this back. `function` is correct here

**Review by RealOrangeOne — CHANGES_REQUESTED** [review]
Overall looks pretty good, a couple small comments. Amazing work! 🎉

File: tests/integration/apis/Easing.test.js (hunk 2 above)
Comment (dusave) [reply tests/integration/apis/Easing.test.js:17]: Arrow functions should be just fine. You're using them above on line 14 already. Is there something I'm missing?

File: .travis.yml (hunk 1 above)
Comment (dustinsavery) [reply .travis.yml]: Sure, I can add that

File: tests/integration/apis/Easing.test.js (hunk 2 above)
Comment (jasonfma) [reply tests/integration/apis/Easing.test.js:17]: Sorry, this change was part of the eslint rule update. I just wholesale applied the rule fixes. I can revert, but do you have more context as to why this one needs to be function?

File: .travis.yml (hunk 1 above)
Comment (jasonfma) [reply .travis.yml]: Should we use `8.9.1` as that is the LTS version?

**Comment by jasonfma** [comment]
@RealOrangeOne just to be clear, this is a PR into the `rewrite` branch that you created. This is not a merge into master.

File: tests/integration/apis/Easing.test.js (hunk 2 above)
Comment (RealOrangeOne) [reply tests/integration/apis/Easing.test.js:17]: I assumed all were using `function`, in that case this is totally fine

File: .travis.yml (hunk 1 above)
Comment (RealOrangeOne) [reply .travis.yml]: Using `8` specifically targets the latest release at build time, which is fine. I quite like the idea of testing against the base version as well

**Comment by jasonfma** [comment]
@RealOrangeOne Looks like node 8.0.0 is broken. I'm going to set to 8.9.1 since it's the first LTS version.

**Comment by RealOrangeOne** [comment]
Just seen the build log, agreed that's very weird. I think sticking to that version will probably be fine, it's early enough!

**Comment by jasonfma** [comment]
@RealOrangeOne can you rebuild? I think it failed because of https://github.com/npm/registry/issues/255

**Comment by jasonfma** [comment]
> dustinsavery commented on Nov 9, 2017
> Should probably add react, react-dom, and react-native as dev dependencies also. When trying to build locally without those packages, it fails

I think this is set up this way so the build can test with various versions of each. I did add them as `peerDeps` so you can see the warnings.

@RealOrangeOne comments addressed

**Comment by jasonfma** [comment]
@RealOrangeOne will you have some time soon to take a look? I'm happy to help here where I can too. This PR is to merge into your existing `rewrite` branch #120 where there are still some merge conflicts to resolve as well - so this is not final or going into `master`.

**Comment by RealOrangeOne** [comment]
All these changes look good to me! Let's get this into `rewrite` and see how it affects it!

**Review by RealOrangeOne — APPROVED** [review]

**Comment by lalop** [comment]
I tried rewrite branch but I get this error 
```
Cannot find module './build/react-native-mock.js' from 'mock.js'
```

Will we use this branch as master ?

**Comment by RealOrangeOne** [comment]
If you're downloading straight from GitHub, you'll have to build the compiled code to actually use it. 

Yes, this code will eventually be pushed into master. Time frame unknown, but shouldn't be too long!

**Comment by lalop** [comment]
I get it via yarn, but build folder isn't present.
Is that normal ?

**Comment by RealOrangeOne** [comment]
Yarn, npm, makes no difference. We don't commit the `build/` directory for obviousl reasons, committing build artifacts is bad practice. If you want to use a version from somewhere other than npm, you'll need to run the build yourself from the project directory, else you'll only have the source files, which won't work.

**Comment by lalop** [comment]
I'm agree with you but I'm not sure to understand.
I get react-native-mock via yarn, that provide source code not built but https://github.com/RealOrangeOne/react-native-mock/blob/rewrite/mock.js call `./build/react-native-mock.js` so is there something wrong here or should I build the package even if I get it via npm/yarn ?

**Comment by RealOrangeOne** [comment]
hmm, this sounds odd, and probably not the best place to discuss. Perhaps raise an actual issue so it can be looked into properly

**Comment by lalop** [comment]
done https://github.com/RealOrangeOne/react-native-mock/issues/157

**Pull request merged**